openmls/src/framing/validation.rs

```diff
@@ -0,0 +1,266 @@
+    credential: Option<Credential>,
+    aad_option: Option<Vec<u8>>,
+}
+
+impl UnverifiedMessage {
+    pub(crate) fn from_decrypted_message(
+        decrypted_message: DecryptedMessage,
+        credential: Option<Credential>,
+    ) -> Self {
+        UnverifiedMessage {
+            plaintext: decrypted_message.plaintext,
+            credential,
+            aad_option: None,
+        }
+    }
+
+    pub fn aad(&self) -> &Option<Vec<u8>> {
+        &self.aad_option
+    }
+    pub fn sender(&self) -> &Sender {
+        self.plaintext.sender()
+    }
+    pub fn credential(&self) -> Option<&Credential> {
+        self.credential.as_ref()
+    }
+    pub(crate) fn into_parts(self) -> (VerifiableMlsPlaintext, Option<Credential>) {
+        (self.plaintext, self.credential)
+    }
+}
+
+pub enum UnverifiedContextMessage {
+    Member(UnverifiedMemberMessage),
+    External(UnverifiedExternalMessage),
+}
+
+impl UnverifiedContextMessage {
+    pub(crate) fn from_unverified_message_with_context(
+        unverified_message: UnverifiedMessage,
+        serialized_context: Vec<u8>,
+        membership_key: &MembershipKey,
+        backend: &impl OpenMlsCryptoProvider,
+    ) -> Result<Self, ValidationError> {
+        // Decompose UnverifiedMessage
+        let (mut plaintext, credential_option) = unverified_message.into_parts();
+
+        if plaintext.sender().is_member() {
+            // Verify the membership tag
+            if plaintext.wire_format() != WireFormat::MlsCiphertext {
+                // Add serialized context to plaintext
+                plaintext.set_context(serialized_context);
+                // Verify the membership tag
+                plaintext.verify_membership(backend, membership_key)?;
+            }
+        }
+        match plaintext.sender().sender_type {
+            SenderType::Member => {
+                if let Some(credential) = credential_option {
+                    Ok(UnverifiedContextMessage::Member(UnverifiedMemberMessage {
+                        plaintext,
+                        credential,
+                    }))
+                } else {
+                    // If the sender is a member, there must be a credential
+                    Err(ValidationError::LibraryError)
+                }
+            }
+            // TODO #192: We don't support external senders yet
+            SenderType::Preconfigured => todo!(),
+            SenderType::NewMember => todo!(),
+        }
+    }
+}
+
+pub struct UnverifiedMemberMessage {
+    plaintext: VerifiableMlsPlaintext,
+    credential: Credential,
+}
+
+impl UnverifiedMemberMessage {
+    pub(crate) fn into_verified(
+        self,
+        backend: &impl OpenMlsCryptoProvider,
+        signature_key: Option<&SignaturePublicKey>,
+    ) -> Result<VerifiedMemberMessage, ValidationError> {
+        // If a signature key is provided it will be used,
+        // otherwise we take the key from the credential
+        match signature_key {
```

Comment:
Sorry to nitpick here, but all this matching seems a bit much. Maybe something like this?
```
        let verified_member_message = if let Some(signature_public_key) = signature_key {
            self.plaintext
                .verify_with_key(backend, signature_public_key)
        } else {
            self.plaintext.verify(backend, &self.credential)
        }
        .map(|plaintext| VerifiedMemberMessage { plaintext })?;
        Ok(verified_member_message)
```